File: scripts/_events.py
"""_events.py -- shared structural emit helper for framework events (#635).

Behavioral events vBRIEF
(``vbrief/proposed/2026-04-27-635-events-behavioral-wiring.vbrief.json``)
lands the 3 behavioral events: ``session:interrupted`` /
``session:resumed`` (paired), ``plan:approved``, and ``legacy:detected``.
The detection-bound sibling vBRIEF lands its 5 events on top of the same
helper.

Why this lives in ``scripts/_events.py``: per the canonical #642 workflow
comment locked decisions, framework events are a STRUCTURAL artifact, not
prose. The emit helper + JSONL append-only log are the deterministic
encoding form; the registries under ``events/*.yaml`` describe the
contract.

Storage
-------
Events are appended to ``.deft/events.jsonl`` under the project root.
``.deft/`` is gitignored as "Generated Deft exports" (#401 trim arc); the
file is project-local and ephemeral. Tests inject a temp path to keep
tests hermetic.

Pairing
-------
``session:interrupted`` / ``session:resumed`` MUST be co-emitted: every
resumed event carries an ``interrupted_id`` referencing the open
interrupt. ``validate_pairing`` returns the list of orphan resumed
records; an empty list means the pair invariant holds. The vBRIEF
acceptance criterion (3) is enforced via this helper.

CLI
---
Agents emit by invoking::

    python -m scripts._events emit <name> --payload '<json>'
    python -m scripts._events emit session:interrupted \
        --session-id <s> --reason context-window-shift
    python -m scripts._events emit session:resumed \
        --session-id <s> --interrupted-id <id>
    python -m scripts._events emit plan:approved \
        --plan-ref <url> --approver <login> --approval-phrase yes
    python -m scripts._events validate-pairing

Handlers
--------
v0 of this surface is emit-only. The behavioral vBRIEF explicitly defers
downstream consumer effects (what plan:approved triggers, what action
legacy:detected drives) to follow-up work. This module exposes
``read_events`` so consumers can be written in subsequent PRs without
churning the emit contract.

Issue: #635 (epic), #642 (workflow umbrella), #634 (determinism ladder).
"""

from __future__ import annotations

import argparse
import json
import os
import secrets
import sys
import time
from collections.abc import Iterable
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

# Default event log location (project-local, gitignored).
DEFAULT_EVENT_LOG: Path = Path(".deft") / "events.jsonl"

# Registered event names. Kept in code so the helper can validate at emit
# time without a YAML loader dependency. The canonical contract still
# lives in ``events/behavioral.yaml`` (and the future
# ``events/detection-bound.yaml`` from the sibling vBRIEF).
KNOWN_EVENTS: frozenset[str] = frozenset({
    "session:interrupted",
    "session:resumed",
    "plan:approved",
    "legacy:detected",
})

# Per-event required payload fields. Mirrors events/behavioral.yaml.
REQUIRED_PAYLOAD: dict[str, tuple[str, ...]] = {
    "session:interrupted": ("session_id", "reason"),
    "session:resumed": ("session_id", "interrupted_id"),
    "plan:approved": ("plan_ref", "approver"),
    "legacy:detected": ("title", "source", "range", "size_bytes"),
}


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def _resolve_log_path(log_path: Path | str | None) -> Path:
    if log_path is not None:
        return Path(log_path)
    env_path = os.environ.get("DEFT_EVENT_LOG")
    if env_path:
        return Path(env_path)
    return DEFAULT_EVENT_LOG


def _new_event_id() -> str:
    """Return a sortable, collision-resistant event id.

    Format: ``<unix-ns>-<8-hex-rand>``. Sortable by emission time so the
    log can be diffed cleanly; rand suffix prevents collision when two
    emissions land in the same nanosecond (rare but possible on Windows).
    """
    return f"{time.time_ns()}-{secrets.token_hex(4)}"


def emit(
    name: str,
    payload: dict[str, Any],
    *,
    log_path: Path | str | None = None,
    detected_at: str | None = None,
) -> dict[str, Any]:
    """Append an event record to the JSONL log and return it.

    ``name`` MUST be a registered event from ``KNOWN_EVENTS``. ``payload``
    MUST contain the required fields for that event per
    ``REQUIRED_PAYLOAD``. ``detected_at`` defaults to the current UTC
    timestamp (ISO 8601, second precision).

    Raises ``ValueError`` for an unknown event name or a missing required
    payload field.
    """
    if name not in KNOWN_EVENTS:
        raise ValueError(
            f"unknown event {name!r}; expected one of "
            f"{sorted(KNOWN_EVENTS)}"
        )
    required = REQUIRED_PAYLOAD.get(name, ())
    missing = [k for k in required if k not in payload]
    if missing:
        raise ValueError(
            f"event {name!r} payload missing required fields: {missing}"
        )

    record: dict[str, Any] = {
        "event": name,
        "id": _new_event_id(),
        "detected_at": detected_at
        or datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "payload": dict(payload),
    }

    target = _resolve_log_path(log_path)
    target.parent.mkdir(parents=True, exist_ok=True)
    line = json.dumps(record, ensure_ascii=False, sort_keys=True)
    with target.open("a", encoding="utf-8") as fh:
        fh.write(line)
        fh.write("\n")
    return record


def read_events(
    log_path: Path | str | None = None,
) -> list[dict[str, Any]]:
    """Return all events from the log in emission order.

    Missing log returns an empty list. Malformed lines are skipped (the
    log is append-only; a partial last line could survive a crash).
    """
    target = _resolve_log_path(log_path)
    if not target.exists():
        return []
    out: list[dict[str, Any]] = []
    with target.open("r", encoding="utf-8") as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            try:
                out.append(json.loads(line))
            except json.JSONDecodeError:
                continue
    return out


def validate_pairing(
    events: Iterable[dict[str, Any]] | None = None,
    *,
    log_path: Path | str | None = None,
) -> list[dict[str, Any]]:
    """Return the list of orphan ``session:resumed`` records.

    A ``session:resumed`` is an orphan if its ``interrupted_id`` does not
    match the ``id`` of any prior ``session:interrupted`` in the same
    log. The vBRIEF acceptance criterion (3) requires the test suite to
    assert this returns ``[]`` for any well-formed event stream.

    Pass ``events`` explicitly to validate an in-memory stream; otherwise
    the helper reads the configured log path.
    """
    if events is None:
        events = read_events(log_path=log_path)
    interrupted_ids: set[str] = set()
    orphans: list[dict[str, Any]] = []
    for record in events:
        name = record.get("event")
        if name == "session:interrupted":
            event_id = record.get("id")
            if isinstance(event_id, str):
                interrupted_ids.add(event_id)
        elif name == "session:resumed":
            payload = record.get("payload") or {}
            ref = payload.get("interrupted_id")
            if not isinstance(ref, str) or ref not in interrupted_ids:
                orphans.append(record)
    return orphans


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------


def _parse_payload_args(args: argparse.Namespace) -> dict[str, Any]:
    """Build the payload dict from the parsed CLI args.

    The ``--payload`` JSON arg, when present, is the seed; named flags
    overlay on top so agents can mix structured input with one-off
    fields.
    """
    payload: dict[str, Any] = {}
    if args.payload:
        try:
            data = json.loads(args.payload)
        except json.JSONDecodeError as exc:
            raise SystemExit(f"--payload is not valid JSON: {exc}") from exc
        if not isinstance(data, dict):
            raise SystemExit("--payload must be a JSON object")
        payload.update(data)

    # Named overlays (only set if provided).
    name_to_field = {
        "session_id": args.session_id,
        "reason": args.reason,
        "interrupted_id": args.interrupted_id,
        "plan_ref": args.plan_ref,
        "approver": args.approver,
        "approval_phrase": args.approval_phrase,
        "pr_number": args.pr_number,
        "detail": args.detail,
        "title": args.title,
        "source": args.source,
        "range": args.range_,
        "size_bytes": args.size_bytes,
        "inline": args.inline,
        "sidecar": args.sidecar,
        "flagged": args.flagged,
    }
    for k, v in name_to_field.items():
        if v is not None:
            payload[k] = v
    return payload


def _build_arg_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="python -m scripts._events",
        description="Emit and inspect structural framework events.",
    )
    sub = parser.add_subparsers(dest="cmd", required=True)

    emit_p = sub.add_parser("emit", help="Append an event to the log.")
    emit_p.add_argument("name", choices=sorted(KNOWN_EVENTS))
    emit_p.add_argument("--payload", help="JSON object with the payload.")
    emit_p.add_argument("--log", dest="log", help="Override event log path.")
    # Convenience flags. Using --range- on the dest avoids shadowing the
    # builtin ``range``.
    emit_p.add_argument("--session-id")
    emit_p.add_argument("--reason")
    emit_p.add_argument("--interrupted-id")
    emit_p.add_argument("--plan-ref")
    emit_p.add_argument("--approver")
    emit_p.add_argument("--approval-phrase")
    emit_p.add_argument("--pr-number", type=int)
    emit_p.add_argument("--detail")
    emit_p.add_argument("--title")
    emit_p.add_argument("--source")
    emit_p.add_argument("--range", dest="range_")
    emit_p.add_argument("--size-bytes", type=int)
    emit_p.add_argument(
        "--inline",
        type=lambda s: s.lower() in {"1", "true", "yes"},
        default=None,
    )
    emit_p.add_argument("--sidecar")
    emit_p.add_argument(
        "--flagged",
        type=lambda s: s.lower() in {"1", "true", "yes"},
        default=None,
    )

    list_p = sub.add_parser("list", help="Print events as JSON lines.")
    list_p.add_argument("--log", dest="log", help="Override event log path.")

    pair_p = sub.add_parser(
        "validate-pairing",
        help="Exit non-zero if any session:resumed is orphan.",
    )
    pair_p.add_argument("--log", dest="log", help="Override event log path.")

    return parser


def main(argv: list[str] | None = None) -> int:
    parser = _build_arg_parser()
    args = parser.parse_args(argv)

    if args.cmd == "emit":
        payload = _parse_payload_args(args)
        try:
            record = emit(args.name, payload, log_path=args.log)
        except ValueError as exc:
            sys.stderr.write(f"emit failed: {exc}\n")
            return 2
        sys.stdout.write(json.dumps(record, ensure_ascii=False))
        sys.stdout.write("\n")
        return 0

    if args.cmd == "list":
        for record in read_events(log_path=args.log):
            sys.stdout.write(
                json.dumps(record, ensure_ascii=False, sort_keys=True)
            )
            sys.stdout.write("\n")
        return 0

    if args.cmd == "validate-pairing":
        orphans = validate_pairing(log_path=args.log)
        if orphans:
            sys.stderr.write(
                f"orphan session:resumed records ({len(orphans)}): "
                f"{[r.get('id') for r in orphans]}\n"
            )
            return 1
        sys.stdout.write("ok\n")
        return 0

    parser.print_help()
    return 2


__all__ = [
    "DEFAULT_EVENT_LOG",
    "KNOWN_EVENTS",
    "REQUIRED_PAYLOAD",
    "emit",
    "main",
    "read_events",
    "validate_pairing",
]


if __name__ == "__main__":  # pragma: no cover - thin CLI shim
    raise SystemExit(main())

File: packages/core/src/lifecycle/events.ts
import { randomBytes } from "node:crypto";
import { appendFileSync, existsSync, mkdirSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

/** Default event log location (project-local). */
export const DEFAULT_EVENT_LOG = join(".deft-cache", "events.jsonl");

const BEHAVIORAL_CATEGORY = "behavioral";

const REQUIRED_BEHAVIORAL_PAYLOAD: Readonly<Record<string, readonly string[]>> = {
  "session:interrupted": ["session_id", "reason"],
  "session:resumed": ["session_id", "interrupted_id"],
  "plan:approved": ["plan_ref", "approver"],
  "legacy:detected": ["title", "source", "range", "size_bytes"],
};

let behavioralRegistryCache: {
  names: ReadonlySet<string>;
  required: Readonly<Record<string, readonly string[]>>;
} | null = null;

function defaultRegistryPath(): string {
  const repoRoot = resolve(dirname(fileURLToPath(import.meta.url)), "..", "..", "..", "..");
  return join(repoRoot, "events", "registry.json");
}

function loadBehavioralRegistry(): {
  names: ReadonlySet<string>;
  required: Readonly<Record<string, readonly string[]>>;
} {
  if (behavioralRegistryCache !== null) {
    return behavioralRegistryCache;
  }
  const data = JSON.parse(readFileSync(defaultRegistryPath(), "utf8")) as Record<string, unknown>;
  const events = data.events;
  const behavioralNames = new Set<string>();
  if (Array.isArray(events)) {
    for (const event of events) {
      if (
        typeof event === "object" &&
        event !== null &&
        !Array.isArray(event) &&
        (event as Record<string, unknown>).category === BEHAVIORAL_CATEGORY
      ) {
        const name = (event as Record<string, unknown>).name;
        if (typeof name === "string") {
          behavioralNames.add(name);
        }
      }
    }
  }
  behavioralRegistryCache = {
    names: behavioralNames,
    required: REQUIRED_BEHAVIORAL_PAYLOAD,
  };
  return behavioralRegistryCache;
}

function registeredBehavioralNames(): ReadonlySet<string> {
  return loadBehavioralRegistry().names;
}

function requiredPayloadMap(): Readonly<Record<string, readonly string[]>> {
  return loadBehavioralRegistry().required;
}

/** Reset the in-process registry cache. Used by tests. */
export function clearRegistryCache(): void {
  behavioralRegistryCache = null;
}

class LazyKnownEvents {
  private resolved(): ReadonlySet<string> {
    return registeredBehavioralNames();
  }

  has(item: unknown): boolean {
    return this.resolved().has(item as string);
  }

  [Symbol.iterator](): Iterator<string> {
    return this.resolved()[Symbol.iterator]() as Iterator<string>;
  }

  get size(): number {
    return this.resolved().size;
  }

  equals(other: unknown): boolean {
    if (!(other instanceof Set) && !(other instanceof LazyKnownEvents)) {
      return false;
    }
    const left = [...this.resolved()].sort();
    const right =
      other instanceof LazyKnownEvents ? [...other.resolved()].sort() : [...other].sort();
    return left.length === right.length && left.every((v, i) => v === right[i]);
  }
}

class LazyRequiredPayload {
  private resolved(): Readonly<Record<string, readonly string[]>> {
    return requiredPayloadMap();
  }

  has(item: unknown): boolean {
    if (typeof item !== "string") {
      return false;
    }
    return Object.hasOwn(this.resolved(), item);
  }

  get(key: string): readonly string[] {
    return this.resolved()[key] ?? [];
  }

  entries(): [string, readonly string[]][] {
    return Object.entries(this.resolved());
  }

  keys(): string[] {
    return Object.keys(this.resolved());
  }

  values(): readonly (readonly string[])[] {
    return Object.values(this.resolved());
  }

  get size(): number {
    return Object.keys(this.resolved()).length;
  }
}

/** Lazy proxy for behavioral event names from the unified registry. */
export const KNOWN_EVENTS = new LazyKnownEvents();

/** Lazy proxy for per-event required payload fields. */
export const REQUIRED_PAYLOAD = new LazyRequiredPayload();

function sortKeysDeep(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortKeysDeep);
  }
  if (value !== null && typeof value === "object") {
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      sorted[key] = sortKeysDeep((value as Record<string, unknown>)[key]);
    }
    return sorted;
  }
  return value;
}

function jsonStringifySorted(value: unknown): string {
  return JSON.stringify(sortKeysDeep(value));
}

function resolveLogPath(logPath?: string | null): string {
  if (logPath !== undefined && logPath !== null) {
    return resolve(logPath);
  }
  const envPath = process.env.DEFT_EVENT_LOG;
  if (envPath !== undefined && envPath.length > 0) {
    return resolve(envPath);
  }
  return resolve(DEFAULT_EVENT_LOG);
}

function newEventId(): string {
  const wallNs = BigInt(Date.now()) * 1_000_000n;
  return `${wallNs}-${randomBytes(4).toString("hex")}`;
}

export interface BehavioralEventRecord {
  readonly event: string;
  readonly id: string;
  readonly detected_at: string;
  readonly payload: Record<string, unknown>;
}

/** Append a behavioral event record to the JSONL log and return it. */
export function emit(
  name: string,
  payload: Record<string, unknown>,
  options: { logPath?: string | null; detectedAt?: string | null } = {},
): BehavioralEventRecord {
  const behavioralNames = registeredBehavioralNames();
  if (!behavioralNames.has(name)) {
    throw new Error(
      `unknown event '${name}'; expected one of ${[...behavioralNames]
        .sort()
        .map((n) => `'${n}'`)
        .join(", ")}`,
    );
  }
  const required = requiredPayloadMap()[name] ?? [];
  const missing = required.filter((key) => !(key in payload));
  if (missing.length > 0) {
    throw new Error(`event '${name}' payload missing required fields: ${JSON.stringify(missing)}`);
  }

  const iso = new Date().toISOString();
  const record: BehavioralEventRecord = {
    event: name,
    id: newEventId(),
    detected_at: options.detectedAt ?? `${iso.slice(0, 19)}Z`,
    payload: { ...payload },
  };

  const target = resolveLogPath(options.logPath);
  mkdirSync(dirname(target), { recursive: true });
  appendFileSync(target, `${jsonStringifySorted(record)}\n`, "utf8");
  return record;
}

/** Return all events from the log in emission order. */
export function readEvents(logPath?: string | null): BehavioralEventRecord[] {
  const target = resolveLogPath(logPath);
  if (!existsSync(target)) {
    return [];
  }
  const out: BehavioralEventRecord[] = [];
  const text = readFileSync(target, "utf8");
  for (const line of text.split(/\r?\n/)) {
    const stripped = line.trim();
    if (stripped.length === 0) {
      continue;
    }
    try {
      out.push(JSON.parse(stripped) as BehavioralEventRecord);
    } catch {}
  }
  return out;
}

/** Return orphan session:resumed records. */
export function validatePairing(
  events?: readonly BehavioralEventRecord[] | null,
  options: { logPath?: string | null } = {},
): BehavioralEventRecord[] {
  const stream = events ?? readEvents(options.logPath);
  const openInterrupts = new Set<string>();
  const orphans: BehavioralEventRecord[] = [];
  for (const record of stream) {
    if (record.event === "session:interrupted") {
      if (typeof record.id === "string") {
        openInterrupts.add(record.id);
      }
    } else if (record.event === "session:resumed") {
      const ref = record.payload?.interrupted_id;
      if (typeof ref === "string" && openInterrupts.has(ref)) {
        openInterrupts.delete(ref);
      } else {
        orphans.push(record);
      }
    }
  }
  return orphans;
}

interface ParsedEmitArgs {
  readonly name?: string;
  readonly payload: Record<string, unknown>;
  readonly log?: string;
}

function parseBooleanFlag(value: string): boolean {
  return ["1", "true", "yes"].includes(value.toLowerCase());
}

function parseEmitArgs(args: readonly string[]): ParsedEmitArgs {
  const payload: Record<string, unknown> = {};
  let name: string | undefined;
  let log: string | undefined;
  let i = 0;
  if (args.length > 0 && !args[0]?.startsWith("--")) {
    name = args[0];
    i = 1;
  }
  while (i < args.length) {
    const arg = args[i];
    if (arg === "--payload") {
      const raw = args[i + 1];
      if (raw === undefined) {
        throw new Error("--payload requires a value");
      }
      const data = JSON.parse(raw) as unknown;
      if (typeof data !== "object" || data === null || Array.isArray(data)) {
        throw new Error("--payload must be a JSON object");
      }
      Object.assign(payload, data);
      i += 2;
      continue;
    }
    if (arg === "--log") {
      log = args[i + 1];
      i += 2;
      continue;
    }
    const flagMap: Record<string, string> = {
      "--session-id": "session_id",
      "--reason": "reason",
      "--interrupted-id": "interrupted_id",
      "--plan-ref": "plan_ref",
      "--approver": "approver",
      "--approval-phrase": "approval_phrase",
      "--detail": "detail",
      "--title": "title",
      "--source": "source",
      "--range": "range",
      "--sidecar": "sidecar",
    };
    if (arg === "--pr-number") {
      payload.pr_number = Number.parseInt(args[i + 1] ?? "", 10);
      i += 2;
      continue;
    }
    if (arg === "--size-bytes") {
      payload.size_bytes = Number.parseInt(args[i + 1] ?? "", 10);
      i += 2;
      continue;
    }
    if (arg === "--inline") {
      payload.inline = parseBooleanFlag(args[i + 1] ?? "");
      i += 2;
      continue;
    }
    if (arg === "--flagged") {
      payload.flagged = parseBooleanFlag(args[i + 1] ?? "");
      i += 2;
      continue;
    }
    const field = flagMap[arg ?? ""];
    if (field !== undefined) {
      payload[field] = args[i + 1];
      i += 2;
      continue;
    }
    throw new Error(`unrecognized arguments: ${arg}`);
  }
  return { name, payload, log };
}

/** CLI entrypoint mirroring scripts/_events.py. */
export function main(argv: readonly string[] = process.argv.slice(2)): number {
  if (argv.length === 0) {
    process.stderr.write("usage: events emit|list|validate-pairing ...\n");
    return 2;
  }
  const cmd = argv[0];
  const rest = argv.slice(1);

  if (cmd === "emit") {
    try {
      const parsed = parseEmitArgs(rest);
      if (parsed.name === undefined) {
        process.stderr.write("emit failed: event name required\n");
        return 2;
      }
      const record = emit(parsed.name, parsed.payload, { logPath: parsed.log ?? null });
      process.stdout.write(`${JSON.stringify(record)}\n`);
      return 0;
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      process.stderr.write(`emit failed: ${message}\n`);
      return 2;
    }
  }

  if (cmd === "list") {
    let log: string | undefined;
    for (let i = 0; i < rest.length; i += 1) {
      if (rest[i] === "--log") {
        log = rest[i + 1];
      }
    }
    for (const record of readEvents(log ?? null)) {
      process.stdout.write(`${jsonStringifySorted(record)}\n`);
    }
    return 0;
  }

  if (cmd === "validate-pairing") {
    let log: string | undefined;
    for (let i = 0; i < rest.length; i += 1) {
      if (rest[i] === "--log") {
        log = rest[i + 1];
      }
    }
    const orphans = validatePairing(null, { logPath: log ?? null });
    if (orphans.length > 0) {
      const ids = orphans.map((r) => r.id);
      process.stderr.write(
        `orphan session:resumed records (${orphans.length}): ${JSON.stringify(ids)}\n`,
      );
      return 1;
    }
    process.stdout.write("ok\n");
    return 0;
  }

  process.stderr.write("usage: events emit|list|validate-pairing ...\n");
  return 2;
}
